src/source/src/txn_channel.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use risingwave_common::transaction::transaction_message::TxnMsg;
+use tokio::sync::{mpsc, oneshot, Semaphore};
+
+pub struct PermitValue(u32);
+
+pub struct TxnMsgWithPermits {
+    pub txn_msg: TxnMsg,
+    pub notificator: oneshot::Sender<usize>,
+    pub permit_value: Option<PermitValue>,
+}
+
+/// Create a channel for transaction messages.
+pub fn txn_channel(initial_permits: usize, batched_permits: usize) -> (Sender, Receiver) {
```

Comment:
This channel is functionally different from the exchange one because it is used on the same node. No network communication is needed.

Reply:
I guess there's no need to introduce `batched_permits` here. It's used for reducing the number of messages to add the permits back through network.